Generator: focus-visible nested polyfill support (#113)
AlaskaAirlines/WC-Generator: docs/focus-visible.md

## Design concerns

The `:focus` state is applied to an element when it is clicked or tapped creating a halo effect. Many designs will often want to remove the outline on focus, however this is not an option if your project is required to be accessible. This presents an interesting design challenge with many opportunities.

## Selectors Level 4 specification

In the next selector level specification draft is [9.4. The Focus-Indicated Pseudo-class: :focus-visible](https://drafts.csswg.org/selectors-4/#the-focus-visible-pseudo).

> The `:focus-visible` pseudo-class applies while an element matches the `:focus` pseudo-class and the UA (User Agent) determines via heuristics that the focus should be made evident on the element. (Many browsers show a “focus ring” by default in this case.)

-- [:focus-visible](https://developer.mozilla.org/en-US/docs/Web/CSS/:focus-visible)

The role of this new selector is to allow for developers to build interfaces that allow for differences based on the user’s input modality, e.g. a mouse vs. the keyboard.

## Focus visible browser support

> With [increased browser support](https://caniuse.com/?search=focus-visible) the `:focus-visible` rarely requires polyfill support. Use of this selector is consistent with any CSS pseudo-class.

Reviewer: ```suggestion
With [increased browser support](https://caniuse.com/?search=focus-visible) the `:focus-visible` selector rarely requires polyfill support. Use of this selector is consistent with any CSS pseudo-class.
```

# Accessibility focus visible

One of the core pillars for supporting accessibility (a11y) within a modern web application is through maintaining the state of focus as a user tabs through an interface. The current implementation of this a11y state is via the `:focus` state introduced in CSS Level 2 (Revision 1).

For a11y reasons, according to the [MDN docs](https://developer.mozilla.org/en-US/docs/Web/CSS/:focus#Accessibility_Concerns), it is advised to NEVER remove the browsers default settings without replacing it with a sufficient substitute.

```css
:focus { outline: none; }
```

> Never just remove the focus outline (visible focus indicator) without replacing it with a focus outline that will pass [WCAG 2.1 SC 1.4.11 Non-Text Contrast](https://www.w3.org/WAI/WCAG21/Understanding/non-text-contrast.html)

## Design concerns

> In contrast to the a11y issue, there is typically a design concern with the `:focus` state as an element within a web app has focus regardless of tab, click or tap. As a result of that, unless there is a requirement to maintain a11y the `:focus` state is typically disabled.

Reviewer: Some awkward phrasing here. Maybe something like:

Since the `:focus` state also applies to an element when it is clicked or tapped, design will often want to remove the outline on focus. However, this is not an option if you want your project to be accessible. This presents an interesting design challenge with many opportunities.

For projects that have an a11y requirement, this is not an option. This in itself is a large design challenge and one with many opportunities.

## Selectors Level 4 specification

In the next selector level specification draft is [9.4. The Focus-Indicated Pseudo-class: :focus-visible](https://drafts.csswg.org/selectors-4/#the-focus-visible-pseudo).

> The `:focus-visible` pseudo-class applies while an element matches the `:focus` pseudo-class and the UA (User Agent) determines via heuristics that the focus should be made evident on the element. (Many browsers show a “focus ring” by default in this case.)

-- [:focus-visible](https://developer.mozilla.org/en-US/docs/Web/CSS/:focus-visible)

> The purpose of this new CSS spec is to allow for developers to build interfaces that speaks specifically to their special needs audience while not having to adjust the experience for those who do not require additional interaction feedback. This also allows for designers to design very specific interactions for those who require additional interaction feedback.

Reviewer: "special needs audience" sounds weird.

## Focus visible browser support

With [increased browser support](https://caniuse.com/?search=focus-visible) the `:focus-visible` rarely requires polyfill support. Use of this selector is consistent with any CSS pseudo-class.

```css
:focus-visible {
  background-color: var(--auro-color-border-error-on-light);
  color: var(--auro-color-base-white);
}
```

> Something to remember, browser defaults have a similar outline style with `:focus-visible` as with a standard `:focus` selector.

Reviewer: ```suggestion
Remember that browser defaults have a similar outline style with `:focus-visible` as with a standard `:focus` selector.
```

## Focus visible polyfill

> The Auro WC-Generator supports the focus-visible [polyfill](https://www.npmjs.com/package/focus-visible), based on the proposed CSS pseudo selector, and allows for this functionality for remaining non-supporting browsers.

Reviewer: Should this be pseudo-class instead of pseudo selector?
